Refactor infrastructure as described in astropy-helpers removal APE

astrofrog opened the pull request:
This PR removes astropy-helpers in the core package and update the infrastructure as described in the APE [A roadmap for package infrastructure without astropy-helpers](https://github.com/astropy/astropy-APEs/pull/52). I am putting this up ahead of the coordination meeting in Socorro, to make it easier to see what the changes would be and discuss specific changes.

The main things this does is:

* Remove astropy-helpers as a submodule, along with the ``ah_bootstrap.py``
* Add a new ``pyproject.toml`` file describing the build-time dependencies (see [PEP 517](https://www.python.org/dev/peps/pep-0517/) and [PEP 518](https://www.python.org/dev/peps/pep-0518/))
* Uses an experimental trimmed down fork of astropy-helpers called extension-helpers (developed in a [branch](https://github.com/astropy/astropy-helpers/tree/extension-helpers) of astropy-helpers) which just does the task of collecting extensions
* Move declarations of package data from ``setup_package.py`` files to ``setup.cfg``
* Switch to using environment variables to control whether to compile against external libraries
* Switch to using ``ASTROPY_DEBUG`` as a way to control whether to compile extensions in debug mode
* Use [setuptools_scm](https://pypi.org/project/setuptools-scm/) to generate the version numbers
* Add a tox configuration to make it easy to automate some of the actions (reviving https://github.com/astropy/astropy/pull/8248)
* Update the CI configuration to use tox whenever possible

I've split out some of the changes into individual self-contained PRs to make the reviewing process a little easier. Specifically, the other PRs that should be reviewed first are:

* Declare all package data in setup.cfg instead of in setup_package.py files [#9729]
* Use environment variables to select system libraries [#9730]
* Use environment variable instead of --debug flag to control extension warnings [#9731]

I've also pushed a [4.1.dev](https://github.com/astropy/astropy/releases/tag/v4.1.dev) tag to make sure that setuptools_scm works properly (how this should be done will be detailed in the packaging docs). For anyone reviewing this, I've tried to keep the commits as tidy as possible, and would recommend looking at individual commits in turn.

* [x] Update the developer documentation to reflect the changes here
* [x] Release a new version of pytest-astropy-header (https://github.com/astropy/pytest-astropy-header/pull/11)
* [x] Make sure the CI configuration is equivalent to before as much as possible
* [x] Add back coverage testing and reporting
* [x] Add changelog entry
* [x] Add help messages to setup.py about new way to run tests and build docs
* [x] Update release docs for version number
* [x] Wait until the APE is discussed and approved
* [x] Update the Travis configuration to re-enable cron builds
* [x] Rebase once #9729, #9730 and #9731 are merged
* [x] Finalize https://github.com/astropy/pytest-astropy/pull/29 and release pytest-astropy
* [x] Update CI config to take into account new pytest-astropy release

Things that I won't do as part of this PR and can be done in follow-ups instead:

* Switch to Azure Pipelines or GitHub actions
* Re-implement the equivalent to the ``--readonly`` option (https://github.com/astropy/astropy/issues/9852)

EDIT: Close #9871

Cadair commented on pyproject.toml:
```diff
@@ -0,0 +1,9 @@
+requires = ["setuptools",
+            "setuptools_scm",
+            "wheel",
+            "cython",
```
You should pin Cython to ensure you get consistent source generation.

astrofrog replied:
Done

Cadair commented on setup.py:
```diff
@@ -1,35 +1,18 @@
 # We set up the following variable because we then use this in astropy/__init__.py
 # to make sure that we aren't importing astropy during the setup process (we used
 # to do this)
 builtins._ASTROPY_CORE_SETUP_ = True
```
Do we still need this?

astrofrog replied:
Removed

mhvk commented:
@astrofrog - this is very nice indeed! While I like seeing everything together (and especially that almost everything leads to code reduction rather than increase!!), I do wonder if it isn't best to actually merge roughly following the order you have your commits, so we can discuss smaller pieces in one place. For instance, it would seem that 69df655eab111e6eb55eace43e7a006085e6ab0f, which moves packaging information to `setup.cfg` is super-uncontroversial and could immediately get merged.

As another example, the `--debug` option, which I will admit I had never looked into, seems rather mis-named - in that it seems to change only whether warnings are visibly generated rather than set up the compiled code to be parsed by a debugger. I may well be wrong about this, but think it is easier to discuss other names if we treat it as a separate commit.

astrofrog commented:
@mhvk - I agree that some things here can be broken out into standalone PRs - some of them have to be done together though, e.g. pyproject.toml and tox. I'll take a look tomorrow at what could be split out and merged first.

bsipocz commented:
> Switch to Azure Pipelines

I would rather switch straight to Actions?

astrofrog commented:
> I would rather switch straight to Actions?

Yes we should discuss options next week! But in any case I want to not mix that up with this PR.

pllim commented on pyproject.toml:
```diff
@@ -0,0 +1,9 @@
+            "setuptools_scm",
+            "wheel",
+            "cython==0.29.14",
+            "jinja2==2.10.3",
```
Why the exact pinning here?

Cadair replied:
To ensure that all builds are predictable. It's to replicate the behavior of these files being generated on the machine that builds the sdist.

pllim commented on setup.cfg:
```diff
@@ -1,6 +1,5 @@
 [metadata]
 name = astropy
-version = 4.1.dev
 provides = astropy
```
I thought `provides` is deprecated?

astrofrog replied:
Fixed!

saimn commented on pyproject.toml:
```diff
@@ -0,0 +1,9 @@
+            "wheel",
+            "cython==0.29.14",
+            "jinja2==2.10.3",
+            "oldest-supported-numpy",
```
Nice ! :clap:

astrofrog commented:
Note that the commented out ``DOUBLE_PLAY`` configuration in Travis is because it did not actually work as it should, and I've opened a separate PR to deal with that: https://github.com/astropy/astropy/pull/9805

astrofrog commented:
I believe that this is now ready for review! I'll also send an email to astropy-dev about it.

mhvk commented on .circleci/config.yml:
```diff
@@ -44,7 +38,9 @@ jobs:
-          command: python3 setup.py test -P visualization --remote-data=astropy --open-files -a "--mpl --mpl-results-path=$PWD/results -W ignore:np.asscalar"
+          command: |
+            pip install -e .[test] pytest-mpl
+            pytest astropy/visualization --remote-data=astropy --open-files --mpl --mpl-results-path=$PWD/results -W ignore:np.asscalar
```
I think these tests (and the ones below) now no longer include the documentation. For visualization, that seems a bad idea... (this goes back to @bsipocz's comment that the one single thing to keep from `setup.py test` is the `-P` option!)

astrofrog replied:
As promised, I've now written up a small pytest plugin to add the ``-P`` option:

https://github.com/astropy/pytest-filter-subpackage

I've updated this PR to use this - for now I'm installing the plugin from the git repository, but before this PR is merged, I'll do a release and add it as a pytest-astropy dependency.

mhvk replied:
Super!!

astrofrog replied:
I've now made a pytest-astropy release, and am about to push changes to the CI to use this, so resolving this.

mhvk commented on .travis.yml:
```diff
@@ -18,63 +18,25 @@ stage: Comprehensive tests
     apt:
         packages:
-            - graphviz
             - language-pack-de
```
Looking at this, why is German included here? `language-pack-de` (though arguably not for this PR)

astrofrog replied:
I can't remember, I think some tests may force the locale to be ``DE``. I'll try and double check.

saimn replied:
Yes, some `io.ascii` tests depend on using a difference locale. Because the locale changes how c functions parse strings.

astrofrog replied:
Just for the record, the test this is for is test_non_C_locale_with_fast_reader

astrofrog replied:
I've added a comment in the Travis file (will push shortly)

astrofrog replied:
Done

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+        # the oldest version of all dependencies where this is known.
+        - language: python
+          python: 3.6
+          env: TOXENV="test-oldestdeps"
```
Insert hyphen? (`oldest-deps`) or just remove `deps`? Or in full, `test-oldest-dependencies` or even `test-with-oldest-dependencies`

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+        - language: python
+          python: 3.6
+          env: TOXENV="test-oldestdeps"
+               TOXPOSARGS="--open-files"
```
Shouldn't testing for open files be in the default?

astrofrog replied:
This slows down the running of tests, so we only included it in one build. I'm trying to avoid any actual changes to the test matrix here, focusing on just translating the old one into the new syntax. I suggest we keep any actual configuration changes for a separate PR.

mhvk replied:
Sounds good.  If you have the energy add something like "we also run the (slow) tests for open files here"

astrofrog replied:
Done

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-          env: PYTHON_VERSION=3.7 CONDA_DEPENDENCIES=$CONDA_ALL_DEPENDENCIES
-               PIP_DEPENDENCIES="scikit-image $ASDF_PIP_DEP"
-               SETUP_CMD='test -a "--durations=50"'
+          env: TOXENV="test-alldeps"
```
I'd again include a hyphen (and this probably means that one should not delete `deps`; though perhaps just spell it out, it being the pythonic way not to use abbreviations? I.e., `test-all-dependencies`)

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+        - language: python
+          python: 3.7
+          env: TOXENV="test-alldeps"
+               TOXPOSARGS="--remote-data=astropy"
```
This is the default, no need to repeat.

astrofrog replied:
The default is an empty string, so this is needed.

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
         # Try on Windows
         - os: windows
           stage: Final tests
-          env: PYTHON_VERSION=3.7 SETUP_CMD='test --readonly'
```
`readonly` is gone? It is the default with `tox`?

astrofrog replied:
See #9852

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-               PIP_DEPENDENCIES="scikit-image jplephem $ASDF_PIP_DEP"
+          env: PYTHON_VERSION=3.7
+               TOXENV="test-alldeps"
+               TOXPOSARGS="--remote-data=astropy"
```
`POSARGS` is odd, here. Really these are extra arguments passed on to the actual code doing the work inside the tox environment. Maybe `TOX_EXTRA_ARGS` or (better) `PYTEST_OPTIONS` (since that seems to be all it is used for).

astrofrog replied:
In general I'd be in favor of making this clearer, but since we might one day allow additional arguments to be passed to non-test commands, I don't think ``PYTEST_OPTIONS`` is ideal. In addition once we switch to using https://github.com/astropy/ci-helpers/pull/429 we'll need to use a standard name. The term ``posargs`` has a specific meaning when using tox, but I could see using ``TOX_EXTRA_ARGS`` for readability.

mhvk replied:
If it is a standard environment variable for `tox`, then I'm fine with keeping it. Maybe have a comment on top where those are defined as empty as the default or so?

astrofrog replied:
I've added a comment to explain this

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-               CONDA_DEPENDENCIES="$CONDA_ALL_DEPENDENCIES clang"
-               PIP_DEPENDENCIES="scikit-image jplephem $ASDF_PIP_DEP"
+          env: PYTHON_VERSION=3.7
+               TOXENV="test-alldeps"
```
I'd prefer underscores to keep things more readable, and am generally in favour of spelling things out if possible, so `TOX_ENV` or `TOX_ENVIRONMENT`

saimn replied:
`TOXENV` is tox thing, allowing to select the environment instead of `-e`.

astrofrog replied:
Yes since this has specific meaning I think we should keep it as-is.

mhvk replied:
Like for `TOXPOSARGS`, maybe define an empty default on top and have a comment there about what it does.

astrofrog replied:
I've added a comment at the top

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-               PIP_DEPENDENCIES=$DEV_PIP_DEP
-               MATPLOTLIB_VERSION=dev
+          env: TOXENV="test-devdeps"
+               TOXPOSARGS="--remote-data"
```
Explicit is better than implicit: `--remote-data=all`

astrofrog replied:
Done

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-               CONDA_DEPENDENCIES=''
-               PIP_DEPENDENCIES=$DEV_PIP_DEP
-               MATPLOTLIB_VERSION=dev
+          env: TOXENV="test-devdeps"
```
I'd do `test-dev-dependencies` (even I wouldn't go to `test-development-dependencies`...)

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-        - os: linux
+        - language: python
+          python: 3.7
           stage: Final tests
```
Off-topic, but wonder if one can write two stages here, `Final tests, cron`?

astrofrog replied:
I'm not sure, but I'd rather experiment in another PR, as there is already a lot going on here :)

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+          python: 3.7
+          stage: Final tests
+          env: TOXENV="test-devdeps"
+               TOXPOSARGS="--remote-data"
```
`--remote-data=all`

astrofrog replied:
Done

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+        export TOXPOSARGS=$TOXPOSARGS" --cov-report=xml:"$TRAVIS_BUILD_DIR"/coverage.xml";
+      fi
+
+    - echo "TOXENG="$TOXENG
```
`TOXENV`???

astrofrog replied:
Fixed

mhvk commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+      fi
+
+    - echo "TOXENG="$TOXENG
+    - echo "TOXARGS="$TOXARGS
```
`TOXARGS` are not used anywhere, so I would remove it (arguably the point is to define all those in an environment anyway; but in any case, we shouldn't introduce anything just because it might be handy).

saimn replied:
For debugging it could be handy to have `TOXARGS='-v'` by default ?

astrofrog replied:
I agree with @saimn that ``-v`` would be good to have. We do also need to make sure we use a standardized name here once we use https://github.com/astropy/ci-helpers/pull/429.

mhvk replied:
OK, if we use it for `-v` then it is fine to introduce.

astrofrog replied:
Done

mhvk commented on astropy/io/misc/asdf/tags/coordinates/tests/__init__.py:
```diff
@@ -0,0 +1,6 @@
+import pytest
```
All these changes would seem to be very suitable to be split out... (in case this review becomes long...)

astrofrog replied:
https://github.com/astropy/astropy/pull/9806

astrofrog replied:
Resolving since this was done.

mhvk commented on astropy/io/ascii/setup_package.py:
```diff
@@ -11,6 +13,6 @@ def get_extensions():
     ascii_ext = Extension(
         name="astropy.io.ascii.cparser",
-        include_dirs=["numpy"],
+        include_dirs=[numpy.get_include()],
```
Didn't we do these already? Needs rebase?

astrofrog replied:
No this wasn't done in other PRs since it requires the ``pyproject.toml`` changes.

mhvk replied:
OK, must have seen it here before then.  Anyway, resolved.

mhvk commented on docs/development/codeguide.rst:
```diff
@@ -5,64 +5,51 @@
 
-* All code must be compatible with Python 3.5 and later.
-  Usage of ``six``, ``__future__``, and ``2to3`` is not longer acceptable.
+* All code must be compatible with Python 3.6 and later.
```
Can we refer to astropy's minimal python version or so?

astrofrog replied:
You mean to avoid hard-coding the minimum version? We could link to the ``setup.cfg`` file and point to the ``requires_python`` section?

mhvk replied:
Yes, something like that. I just find that in my own packages I forget to update these parts...

astrofrog replied:
Done

mhvk commented on docs/development/codeguide.rst:
```diff
@@ -5,64 +5,51 @@
+  Usage of ``six``, ``__future__``, and ``2to3`` is no longer acceptable.
 
 * The new Python 3 formatting style should be used (i.e.
   ``"{0:s}".format("spam")`` instead of ``"%s" % "spam"``).
```
Should this be updated to f-strings?

astrofrog replied:
Changed

mhvk commented on docs/development/testguide.rst:
```diff
@@ -30,20 +29,23 @@ accessed through the ``python setup.py test`` command.
+or install the core package in 'editable' mode specifying the ``[test]``
+option::
+
+    pip install .[test]
```
No `-e` here? (I don't understand `pip` much, so may be off base)

astrofrog replied:
Fixed

mhvk commented on docs/development/testguide.rst:
```diff
@@ -67,50 +69,65 @@ options require the `pytest-pep8 plugin
-Turn on PEP8 checking by passing ``--pep8`` to the ``test`` command. This will
-turn off regular testing and enable PEP8 testing.
+The latter is required if you need to test parts of the package that rely
+on certain entry points being installed.
```
This presumes someone knows what "entry points" are - I'm afraid I don't. Is this about the C/Cython extensions?

astrofrog replied:
No, entry points are separate from extensions and are described in https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points - I think the best solution here is probably to just link the term 'entry points' to that setuptools docs page?

mhvk replied:
Yes, the link would solve this.

astrofrog replied:
Done

mhvk commented on docs/development/testguide.rst:
```diff
@@ -67,50 +69,65 @@ options require the `pytest-pep8 plugin
+subpackages.  For example, to run only the ``wcs`` tests from the
+commandline::
+
+    pytest -P wcs
```
If this is possible and does do `docs` too, then use it in circle-ci!!

astrofrog replied:
Done!

mhvk commented on docs/development/testguide.rst:
```diff
@@ -67,50 +69,65 @@ options require the `pytest-pep8 plugin
-This will run all the default tests for Astropy.
+This will run all the default tests for Astropy (but will not run the
+documentation tests in the ``.rst`` documentation since those files are
+not installed.
```
Missing closing parenthesis

astrofrog replied:
Fixed

mhvk commented on docs/development/testguide.rst:
```diff
@@ -143,63 +160,65 @@ Astropy Test Function
+or run the test suite with all optional dependencies with::
 
-Or, to run only the ``wcs`` and ``utils`` tests::
+    tox -e test-all
```
I think in `.travis.yml` this was `test-all-deps`....

astrofrog replied:
Fixed

mhvk commented on docs/development/testguide.rst:
```diff
@@ -143,63 +160,65 @@ Astropy Test Function
+You can see a list of available test environments with::
 
-Or from Python::
+    tox -l
```
I guess we cannot make this a doctest?

astrofrog replied:
I don't think so, so it is a shell command?

astrofrog replied:
I'm going to resolve this since it isn't possible.

mhvk commented on docs/development/testguide.rst:
```diff
@@ -143,63 +160,65 @@ Astropy Test Function
+anything after this will be passed to pytest, e.g.::
 
-    python setup.py test -t units/index.rst
+    tox -e test -- -v --pdb
```
It doesn't really matter, but I suppose you checked that just passing in a filename after doesn't work?

astrofrog replied:
I've rephrased this since it **is** now possible to pass ``-P wcs`` to tox for example

mhvk commented on docs/development/testguide.rst:
```diff
@@ -143,63 +160,65 @@ Astropy Test Function
+Using the :ref:`openfiles-plugin` plugin (which is installed automatically
+when installing pytest-astropy),  we can test whether any of the unit tests
+inadvertently leave any files open.  Since this greatly slows down the time it
+takes to run the tests, it is turned off by default.
```
Ah, now I see why we do not do it by default in travis - sorry!

astrofrog replied:
Ah just saw this comment :) Marking as resolved.

mhvk commented on docs/development/testguide.rst:
```diff
@@ -212,10 +231,10 @@ For more information on the ``pytest-openfiles`` plugin see
-Astropy can use `coverage.py <https://coverage.readthedocs.io/en/latest/>`_ to
-generate test coverage reports.  To generate a test coverage report, use::
+Coverage reports can be generated using the `pytest-cov
+<https://pypi.org/project/pytest-cov/>`_ plugin by using e.g.::
```
Is this a dependency of `pytest-astropy` as well? If so, do mention.

mhvk replied:
Leftover from previous review... Though from `setup.cfg` it is clear it is not a dependency of `pytest-astropy` - find to leave that as is for now.

astrofrog replied:
pytest-astropy 0.8 now includes pytest-cov, so I've updated the wording here to mention that this is the case

mhvk replied:
Super!

mhvk commented on docs/development/testguide.rst:
```diff
@@ -227,38 +246,21 @@ Running tests in parallel
-installed using `pip`_::
-
-    pip install pytest-xdist
+<https://pypi.org/project/pytest-xdist>`_ plugin.
```
Again, if installed by default as a dependency, do mention. (If not, that's OK; coverage I do think should be part of it)

astrofrog replied:
It isn't included in pytest-astropy so I won't mention this

mhvk commented on docs/development/testguide.rst:
```diff
@@ -634,7 +630,7 @@ bundled version using e.g.::
     from ...tests.helper import pytest
 
-but this is no longer the recommended method.
+but this is no longer supported.
```
I'd just remove this whole section about things prior to 2.0 - that's too long ago!

astrofrog replied:
Done

mhvk commented on docs/install.rst:
```diff
@@ -188,37 +195,28 @@ Building from Source
 
 For Debian/Ubuntu::
 
-    sudo apt-get install python3-dev python3-numpy-dev python3-setuptools cython3 python3-jinja2 python3-pytest-astropy
```
I just changed this section to be more `pip` agnostic - many people on distributions prefer not to create another distribution system, especially one that does not clean up (like `pip`) and does not work for all users. Can you leave this as is?

astrofrog replied:
I've changed this back for now

mhvk commented on docs/install.rst:
```diff
@@ -251,23 +249,15 @@ Development Repository
 using this command::
 
-   git clone --recursive git://github.com/astropy/astropy.git
+   git clone git://github.com/astropy/astropy.git
```
So does the version numbering work for users too, without the extra git stuff needed in `.travis.yml`?

astrofrog replied:
Yes, it's just that Travis does a shallow clone by default

mhvk commented on docs/install.rst:
```diff
@@ -470,20 +476,12 @@ can also open issues in the repositories for some of the dependencies:
+is to use `tox <https://tox.readthedocs.io/en/latest/>`_::
 
-    python setup.py test
+    tox -e test-all
```
should this be `test-alldeps` (or `test-all-dependencies`)?

astrofrog replied:
Fixed

mhvk commented on docs/testhelpers.rst:
```diff
@@ -36,8 +36,8 @@ Reference/API
 
-When executing tests with either `astropy.test` or ``python setup.py test`` the
-call to pytest is controlled by the `astropy.tests.runner.TestRunner` class.
+When executing tests with either `astropy.test` the call to pytest is controlled
```
Delete "either"

astrofrog replied:
Fixed

mhvk commented on pyproject.toml:
```diff
@@ -0,0 +1,9 @@
+requires = ["setuptools",
+            "setuptools_scm",
+            "wheel",
+            "cython==0.29.14",
```
Could one do "oldest-supported-cython"? Aside: this will never matter for regular pip install/test/etc., correct? Just for creating the distribution.  Could you add a general comment on line 1 what this file is for?

astrofrog replied:
Unlike Numpy where we want the oldest supported Numpy as a function of Python version, and this should not depend on project, for Cython we want to use the same version for all Python versions and we may need to pin to a very recent version if e.g. we want to use a new feature. I think different projects may want to pin to different versions depending on what they need so it's less clear that a meta-package makes sense here.

mhvk commented on setup.cfg:
```diff
@@ -11,48 +9,50 @@ description = Community-developed python astronomy tools
-    Programming Language :: Python :: Implementation :: CPython
-    Topic :: Scientific/Engineering :: Astronomy
-    Topic :: Scientific/Engineering :: Physics
+	Intended Audience :: Science/Research
```
Why the change in indentation? 4 spaces seems just right...

astrofrog replied:
Fixed

mhvk commented on setup.cfg:
```diff
@@ -11,48 +9,50 @@ description = Community-developed python astronomy tools
+	Topic :: Scientific/Engineering :: Physics
 
 [options]
+packages = find:
```
What is this? Maybe add comment? People will look to this `setup.cfg` for inspiration in affiliated packages... (well, I do)

astrofrog replied:
This will cause all folders containing an ``__init__.py`` to be found. I'll add a comment.

astrofrog replied:
Done

mhvk commented on setup.cfg:
```diff
@@ -11,48 +9,50 @@ description = Community-developed python astronomy tools
-    showtable = astropy.table.scripts.showtable:main
-    volint = astropy.io.votable.volint:main
-    wcslint = astropy.wcs.wcslint:main
+	fits2bitmap = astropy.visualization.scripts.fits2bitmap:main
```
Same, please don't change indentation, just makes review harder!

astrofrog replied:
I'm not sure why this happened, but this has been fixed!

mhvk commented on setup.cfg:
```diff
@@ -11,48 +9,50 @@ description = Community-developed python astronomy tools
     coverage
     skyfield
+cov =
+    pytest-cov
```
As mentioned above, shouldn't this just be part of `pytest-astropy`?

astrofrog replied:
Done!

mhvk commented on setup.py:
```diff
@@ -1,35 +1,62 @@
 #!/usr/bin/env python
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
-# NOTE: The configuration for the package, including the name, version, and
```
Can we keep an edited version of this note?

astrofrog replied:
Done

mhvk commented on setup.py:
```diff
@@ -1,35 +1,62 @@
+directly with::
 
-import ah_bootstrap  # noqa
+    pip install -e .
```
Should this be `-e .[test]`?

astrofrog replied:
Fixed

mhvk commented on setup.py:
```diff
@@ -1,35 +1,62 @@
+
+    tox -e build_docs
+
+If you don't already have tox installed, you can install it with:
```
Maybe give the alternative of
```
pip install -e .[docs]
cd docs
make html
```
here as well?

astrofrog replied:
Added

mhvk commented on setup.py:
```diff
@@ -1,35 +1,62 @@
+    print(DOCS_HELP)
+    sys.exit(1)
+
+# NOTE: The configuration for the package, including the name, version, and
```
Ah, it is here. I would keep it on top.

astrofrog replied:
Fixed

mhvk commented on tox.ini:
```diff
@@ -0,0 +1,81 @@
+    oldestdeps: numpy==1.16.*
+    oldestdeps: matplotlib==2.1.*
+    oldestdeps: asdf==2.4.*
+    numpy116: numpy==1.16.*
```
These explicit versions seem like a bad idea. Why not just explicitly do this where needed?

astrofrog replied:
Because there is no nice way to do this from outside tox (there are ways, but they are ugly). The idea here is to make it so that all test configurations are easy to run either on CI or locally, so that failures in CI can be reproduced locally.

The ``oldestdeps`` is especially useful since there is no other way to make it easy to make sure things always work with the oldest versions of dependencies.

astrofrog replied:
Just to give an example, you could pin numpy from outside tox by doing:

```
tox -e test --pypi-filter='numpy==1.16.*'
```

but it's much nicer to just do:

```
tox -e test-numpy116
```

My philosophy with tox and CI is to have as much as possible defined in ``tox.ini`` and as little as possible in the CI config so that we can easily transition between CI services.

mhvk replied:
OK, I see, if this *replaces* things in `.travis.yml`, that is great (and, indeed, better!).

mhvk commented on tox.ini:
```diff
@@ -0,0 +1,81 @@
+    alldeps: all
+commands =
+    pip freeze
+    test: pytest --pyargs astropy {toxinidir}/docs --cov astropy --cov-config={toxinidir}/astropy/tests/coveragerc {posargs}
```
I'm confused, does this mean coverage is checked all the time?

astrofrog replied:
I've now fixed this so that coverage is only tested if ``cov`` is in the environment name

mhvk commented on tox.ini:
```diff
@@ -0,0 +1,81 @@
+skip_install = true
+description = check code style, e.g. with flake8
+deps = flake8
+commands = flake8 astropy --count --select=E101,W191,W291,W292,W293,W391,E111,E112,E113,E30,E502,E722,E901,E902,E999,F822,F823
```
I think the selection should be taken from/kept inside `setup.cfg`, so that `flake8` picks up on it (and emacs too!)

saimn replied:
This would shadow other useful errors/warnings in your editor. There is an issue open about this, and a consensus on reducing the number of errors/warnings first before moving to excluding instead of selecting the codes.

astrofrog replied:
Yes indeed as @saimn pointed out, this is deliberate and is why we used to define this list in the Travis configuration. The real fix here is to simply fix all flake8 warnings in the astropy code base so that things just work with the bare flake8 command.

astrofrog replied:
I've added a comment to explain this in tox.ini

mhvk commented on tox.ini:
```diff
@@ -0,0 +1,81 @@
+[tox]
```
General on `tox.ini` - ideally, all actual options for testing/style are kept in `setup.cfg` and this does nothing more than define the tox environments that run them automatically. Also, comments on what things do would be most appreciated! Much of it is not obvious...

astrofrog replied:
I will try and add more comments to the tox.ini file.

astrofrog replied:
I've now added a lot more descriptive comments in the tox file

saimn commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
-          $MAIN_CMD $SETUP_CMD;
-      fi
+    - pip install tox
+    - tox -e $TOXENV $TOXARGS -- $TOXPOSARGS
```
`-e $TOXENV` is probably useless here since tox will read the env var by itself.

astrofrog replied:
Fixed

saimn commented on tox.ini:
```diff
@@ -0,0 +1,81 @@
+    pip freeze
+    test: pytest --pyargs astropy {toxinidir}/docs --cov astropy --cov-config={toxinidir}/astropy/tests/coveragerc {posargs}
+    build_docs: sphinx-build -W -b html . _build/html {posargs}
+    linkcheck: sphinx-build -W -b linkcheck . _build/html {posargs}
```
For readability I would put `build_docs` and `linkcheck` in their own section, otherwise it's a bit difficult to find all the pieces. Something like
```
 [testenv:build_docs]
changedir = docs
description = invoke sphinx-build to build the HTML docs
extras = docs
commands = sphinx-build -W -b html . _build/html {posargs}
```
is much clearer imo, and easier to copy/reuse/modify.

astrofrog replied:
Done

saimn commented on .travis.yml:
```diff
@@ -91,160 +53,150 @@ matrix:
+               TOXPOSARGS="--remote-data=astropy"
+               CONDA_DEPENDENCIES="clang"
                CCOMPILER=clang
+               USE_CI_HELPERS=True
```
I was wondering about the status of ci-helpers with APE17. For what ci-helpers is still useful ?

saimn replied:
It could be for conda installation, as conda is defined just above, but I don't see tox-conda used in the `tox.ini`. Is it planned ?

bsipocz replied:
see https://github.com/astropy/ci-helpers/pull/429

bsipocz replied:
basically we would keep it to be able to centrally limit versions if necessary (let's say we know that a certain pytest, etc is incompatible)

astrofrog replied:
In this case it is still useful to set up different Python versions on non-Linux builds on Travis. We wouldn't need it if we switched to e.g. Azure for example, but that is beyond the scope of this PR.

astrofrog replied:
The issue with using tox-conda is that we would need https://github.com/tox-dev/tox-conda/issues/36 to be implemented properly to be able to use it. So instead here we never use tox-conda but do use conda on Travis to at least set up the Python environment. To simplify things I've now removed ``USE_CI_HELPERS`` and just decide whether to install conda based on the OS.

pllim commented on docs/development/releasing.rst:
```diff
@@ -257,18 +212,11 @@ clean-up tasks to finalize the process.
-      $ git commit -m "Back to development: v<next_version>.dev"
-
-#. Also update the ``CHANGES.rst`` file with a new section for the next version.
+#. Go back to release branch (e.g., ``1.2.x``) and update the ``CHANGES.rst``
```
Also need to tag the dev version here, e.g., v1.2.3.dev, for `setuptools_scm` to pick it up.

astrofrog replied:
This is already mentioned in the section 'Performing a Feature Freeze/Branching new Major Versions' - it is **not** needed inside release branches - that is, if you tag v1.2.2, setuptools_scm will automatically pick that up and call the version v1.2.3.dev in subsequent commits to the branch. The issue that requires adding .dev tags is *only* when doing branching from master.

astrofrog commented on astropy/wcs/setup_package.py:
```diff
@@ -226,7 +227,7 @@ def get_wcslib_cfg(cfg, wcslib_files, include_paths):
 
     # Squelch a few compilation warnings in WCSLIB
-    if setup_helpers.get_compiler_option() in ('unix', 'mingw32'):
+    if get_compiler() in ('unix', 'mingw32'):
```
This is not trivial as illustrated by the implementation of ``get_compiler`` in extension-helpers - we can easily find the compiler used to build Python with the ``ccompiler`` module but that doesn't take into account if the compiler is customized with the ``--compiler`` option to ``setup.py build``.

mhvk replied:
OK, pity...

astrofrog commented on astropy/utils/xml/setup_package.py:
```diff
@@ -1,19 +1,20 @@
 import sys
 
-from astropy_helpers import setup_helpers
+from extension_helpers import pkg_config
```
Fixed

astrofrog commented:
@mhvk - thanks for the detailed review! I think I've addressed most of your comments. The main one I haven't is related to the tox environment names.

Tox environment names are constructed using words ('factors' in tox terminology) separated by hyphens. Hyphens are syntactically meaningful because tox will split up the environment name using the hyphens, then match it against any conditions in the tox file (here is some more [information about tox factors](https://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings)). What this means in practice is that we shouldn't do things like ``test-all-dependencies`` because ``all`` and ``dependencies`` are then treated as separate 'factors' which might get confusing in some cases. In addition, we should avoid being too verbose since these are commands we might have to type semi-regularly, so e.g. ``tox -e test-devdeps`` is easier to type than ``tox -e test-withalldeveloperdependencies``, and environment names can already get long-ish if you combine things, e.g. ``test-devdeps-cov``.

I've now tried to properly document all the existing factors, and you can use ``tox -lv`` (list and verbose) to show the available environments along with a verbose description of what they do:

```
% tox -l -v
using tox.ini: /Users/tom/Dropbox/Code/Astropy/astropy/tox.ini (pid 32231)
using tox-3.14.2 from /Users/tom/python/dev/lib/python3.8/site-packages/tox/__init__.py (pid 32231)
default environments:
py36-test                -> run tests
py36-test-cov            -> run tests and test coverage
py36-test-alldeps        -> run tests with all optional dependencies
py36-test-alldeps-cov    -> run tests with all optional dependencies and test coverage
py36-test-oldestdeps     -> run tests with the oldest supported version of key dependencies
py36-test-oldestdeps-cov -> run tests with the oldest supported version of key dependencies and test coverage
py36-test-devdeps        -> run tests with the latest developer version of key dependencies
py36-test-devdeps-cov    -> run tests with the latest developer version of key dependencies and test coverage
py36-test-numpy116       -> run tests with numpy 1.16.*
py36-test-numpy116-cov   -> run tests and test coverage with numpy 1.16.*
py36-test-numpy117       -> run tests with numpy 1.16.*
py36-test-numpy117-cov   -> run tests and test coverage with numpy 1.16.*
py36-test-numpy118       -> run tests with numpy 1.16.*
py36-test-numpy118-cov   -> run tests and test coverage with numpy 1.16.*
py37-test                -> run tests
py37-test-cov            -> run tests and test coverage
py37-test-alldeps        -> run tests with all optional dependencies
py37-test-alldeps-cov    -> run tests with all optional dependencies and test coverage
py37-test-oldestdeps     -> run tests with the oldest supported version of key dependencies
py37-test-oldestdeps-cov -> run tests with the oldest supported version of key dependencies and test coverage
py37-test-devdeps        -> run tests with the latest developer version of key dependencies
py37-test-devdeps-cov    -> run tests with the latest developer version of key dependencies and test coverage
py37-test-numpy116       -> run tests with numpy 1.16.*
py37-test-numpy116-cov   -> run tests and test coverage with numpy 1.16.*
py37-test-numpy117       -> run tests with numpy 1.16.*
py37-test-numpy117-cov   -> run tests and test coverage with numpy 1.16.*
py37-test-numpy118       -> run tests with numpy 1.16.*
py37-test-numpy118-cov   -> run tests and test coverage with numpy 1.16.*
py38-test                -> run tests
py38-test-cov            -> run tests and test coverage
py38-test-alldeps        -> run tests with all optional dependencies
py38-test-alldeps-cov    -> run tests with all optional dependencies and test coverage
py38-test-oldestdeps     -> run tests with the oldest supported version of key dependencies
py38-test-oldestdeps-cov -> run tests with the oldest supported version of key dependencies and test coverage
py38-test-devdeps        -> run tests with the latest developer version of key dependencies
py38-test-devdeps-cov    -> run tests with the latest developer version of key dependencies and test coverage
py38-test-numpy116       -> run tests with numpy 1.16.*
py38-test-numpy116-cov   -> run tests and test coverage with numpy 1.16.*
py38-test-numpy117       -> run tests with numpy 1.17.*
py38-test-numpy117-cov   -> run tests and test coverage with numpy 1.17.*
py38-test-numpy118       -> run tests with numpy 1.18.*
py38-test-numpy118-cov   -> run tests and test coverage with numpy 1.18.*
build_docs               -> invoke sphinx-build to build the HTML docs
linkcheck                -> check the links in the HTML docs
codestyle                -> check code style, e.g. with flake8
```

I've also updated the developer docs to mention this command (in addition to commenting the tox file extensively). Given this, would you be ok with the more concise factor names as above?

mhvk commented:
@astrofrog - that's a very neat feature of the tox environments! And, yes, given that, short names definitely make sense. I'll have another look at the whole thing now.

mhvk commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
     include:
-        # Try MacOS X.
+
+      # Try MacOS X.
```
Indentation?

astrofrog replied:
Fixed

mhvk commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
                CCOMPILER=clang
                EVENT_TYPE='cron'
-               DOUBLE_PLAY=True
+               # DOUBLE_PLAY=True
```
Still work in progress for `DOUBLE_PLAY`?

astrofrog replied:
Fixed in https://github.com/astropy/astropy/pull/9805 and am now rebasing this

mhvk commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
+          env: TOXENV="test-oldestdeps"
+               TOXPOSARGS="--open-files"
+
+        # Now try with all optional dependencies.
```
Add "and determine coverage."

astrofrog replied:
The coverage test is the next one, and the comment already mentions coverage

mhvk commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
 
 before_install:
 
+    # We need a full clone to make sure setuptools_scm works properly
```
For another PR, but it would seem that logically `setuptools_scm` would need to know only the last tag that this branch is on...

astrofrog replied:
I agree - and indeed probably easiest handled in a separate PR.

mhvk commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
-    - if [[ $SETUP_CMD == *--coverage* ]]; then
-        codecov --gcov-glob "*cextern*";
-      fi
+    - pip install codecov
```
Why no longer inside an `if [[ $TOXENV == *-cov ]]; then`?

astrofrog replied:
Fixed

mhvk commented on docs/development/ccython.rst:
```diff
@@ -13,15 +13,13 @@ subpackage designer, and can be customized however is relevant for the
-that this allows disabling a Cython file by providing an extension that
-includes the Cython file, but giving it the special ``name`` 'cython_skip'. Any
-extension with this package name will not be built by ``setup.py``.
+included in the list of extensions automatically generated extensions.
```
Extraneous "extensions"

saimn commented:
To shorten a bit more env names, it seems the `-test` factor is useless could be removed ? I guess you added this to make it explicit that these are test environments, but to me at least it seems not necessary.

mhvk commented on docs/development/testguide.rst:
```diff
@@ -227,38 +251,21 @@ Running tests in parallel
-directory be set as read-only to ensure the tests are not writing to that
-location.  This mode can be triggered by running the tests like so::
-
-    python setup.py test --readonly
```
Is this automatically done by `tox`?

astrofrog replied:
No, this is the one thing that hasn't been translated to the new infrastructure yet. I'll have a think about it.

mhvk replied:
OK, no worries - I'd just open a new issue about it, so we can go towards merging this!

astrofrog replied:
See #9852

mhvk commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
+          env: TOXENV="test-alldeps-cov"
+               TOXPOSARGS="--remote-data=astropy"
                LC_CTYPE=C.ascii LC_ALL=C
                CFLAGS='--coverage -fno-inline-functions -O0'
```
Is this still necessary? In `tox.ini`, it would seem this gets done automatically...

astrofrog replied:
Indeed, my bad! I added it to tox.ini without removing it here... Fixed now.

mhvk commented:
As long as a default `test` environment does exist!

saimn commented on .travis.yml:
```diff
@@ -92,151 +63,149 @@ matrix:
-          env: PYTHON_VERSION=3.7 CONDA_DEPENDENCIES=$CONDA_ALL_DEPENDENCIES
-               PIP_DEPENDENCIES="scikit-image $ASDF_PIP_DEP"
-               SETUP_CMD='test -a "--durations=50"'
+          env: TOXENV="test-alldeps"
```
So tox will match this to `py36-test-alldeps` based on the python version ? I did not know this was possible. This may be a bit confusing, it's convenient to avoid repeating the python version but I think it would be better if TOXENV was set to the real env name.

astrofrog replied:
Yes if py?? isn't specified it just uses the current default version. But I agree it would be cleaner to be explicit.

astrofrog replied:
Fixed - all tox environments in CI will now have the Python version explicitly specified

astrofrog commented:
Currently the ``test`` part of the tox environment name is optional, but I'd rather keep it explicit in the CI. But I think locally you should be able to do ``tox -e py37`` and it will mean the same as ``tox -e py37-test``.

astrofrog commented:
@mhvk - the main item from your review I can't address at the moment is the ``--readonly option`` to ``test``. I added that option in https://github.com/astropy/astropy/pull/7598 to try and prevent future issues similar to that which @olebole reported in https://github.com/astropy/astropy/issues/7479 where during Debian testing, it was found that some astropy tests were writing to the installed package directory. I'm not sure what the best way is to re-implement this, though I have some ideas. However, since the Debian testing would catch this kind of issue, and in the interest of moving this PR forward, would it be ok for me to open an issue and work on this in a follow-up PR?

mhvk commented:
@astrofrog - yes, absolutely fine to just open a new issue for `--readonly`.

astrofrog commented:
@mhvk  - ok, thanks! I opened an issue here: https://github.com/astropy/astropy/issues/9852

astrofrog commented:
Before this PR is considered final, we will need to merge in https://github.com/astropy/pytest-astropy/pull/29 which adds pytest-filter-subpackage and pytest-cov as dependencies to pytest-astropy, do a new release of pytest-astropy, and update this PR.

astrofrog commented:
@bsipocz @eteq - you had concerns about what the jobs would look like in Travis so you might be interested in one of the commits I just pushed, which adds names to the Travis jobs, which now look like:

<img width="789" alt="Screenshot 2020-01-16 at 11 03 18" src="https://user-images.githubusercontent.com/314716/72519757-f3ad1580-384f-11ea-9ffc-c7a5da83b243.png">

which is a little more readable than our current:

<img width="823" alt="Screenshot 2020-01-16 at 11 05 05" src="https://user-images.githubusercontent.com/314716/72519840-16d7c500-3850-11ea-8239-ea27ff38c5ab.png">

Cadair commented on CHANGES.rst:
```diff
@@ -268,6 +268,19 @@ Other Changes and Additions
+  manage the version number, and adding a ``pyproject.toml`` to opt in to
+  isolated builds as described in PEP 517/518. [#9726]
+
+- The minimum supported version of asdf has been updated to 2.4. [#9726]
```
Is this supposed to be here?

astrofrog replied:
No, it's left over from code I moved to a separate PR - removed!

Cadair commented on docs/development/releasing.rst:
```diff
@@ -64,36 +63,19 @@ packages that use the full bugfix/maintenance branch approach.)
-   necessary if you know your normal python environment has what you need, but
-   you might want to do something like this for safety's sake::
+#. To build the source distribution in an isolated environment and make sure you
+   have all the dependencies required for it, install the `pep517
```
You don't actually need all the deps right? `pep517` will install them for you?

astrofrog replied:
Indeed - this sentence doesn't say you need to install them yourself though currently? I was trying to imply that pep517 took care of that. Should I rephrase?

pllim replied:
I think it is okay to keep as-is.

astrofrog commented:
@mwcraig - you are correct that the version number gets frozen currently when pip is run, however this can be solved by customizing the template for ``version.py``. To keep things simple, I'd like to defer this to a separate PR though - I'll open an issue about it.

taldcroft commented:
One minor niggle is that with `pip install -e .` you get the installed version directory in the source and then an IDE (like VS code) starts picking up files in that directory as project source files.  Any thoughts on this?  I thought about maybe adding something like `/astropy-*` to `.gitignore`?

astrofrog commented:
@mwcraig @Cadair - see the latest commit I pushed (https://github.com/astropy/astropy/pull/9726/commits/4ca3550b8cfc591d3fa12cbbc3d7b0f8ea558a3b) - this will now dynamically update the version if setuptools_scm is installed, and will fall back to the last pip install version if not. It seems to work fine for me. Does this seem reasonable?

astrofrog commented:
@taldcroft - weirdly I don't see the folder you mention. What is the exact path of the folders that are added?

mwcraig commented:
> this will now dynamically update the version if setuptools_scm is installed, and will fall back to the last pip install version if not. It seems to work fine for me. Does this seem reasonable? 

Excellent, @astrofrog! I likely won't have a chance to test this out today; I think the slight performance penalty is worth it to ensure devs see the correct version fi they run the tests via pytest or some other mechanism that skips running setup.

taldcroft commented:
>  weirdly I don't see the folder you mention. What is the exact path of the folders that are added?

I had done this previously on master, not this branch.  Now on this branch I don't see that directory showing up and `pip install -e .` looks clean.  So looks OK!

taldcroft commented:
> this will now dynamically update the version if setuptools_scm is installed, and will fall back to the last pip install version if not. It seems to work fine for me. Does this seem reasonable?

I just tested for both cases (with/without setuptools_scm) and got the expected results.  Fantastic!  It's quite verbose, but getting a dirty version number like `astropy=4.1.dev476+gaa2a02306.d20200116` after making local changes is *really* useful. (I recently switched to zsh + OhMyZsh, which also helps alot).

pllim commented on tox.ini:
```diff
@@ -0,0 +1,117 @@
+commands =
+    pip freeze
+    !cov: pytest --pyargs astropy {toxinidir}/docs {posargs}
+    cov: pytest --pyargs astropy {toxinidir}/docs --cov astropy --cov-config={toxinidir}/.coveragerc {posargs}
```
Question: If we include `.coveragerc` file in `package_data`, does that mean we don't need to specify `--cov-config` anymore?

astrofrog replied:
We still need to specify ``--cov-config`` in all cases because pytest isn't run at the root of the repository.

pllim commented on docs/development/codeguide.rst:
```diff
@@ -295,17 +297,18 @@ Including C Code
-  library in place of the library included in Astropy.
+  library in place of the library included in Astropy, and a user installing
+  the package should be able to opt-in to using the system version using
+  a ``ASTROPY_USE_SYSTEM_???`` environment variable, where ``???`` is the name
```
Can you please provide examples of what `???` can possibly be in real life?

astrofrog replied:
Done

pllim commented on docs/development/testguide.rst:
```diff
@@ -67,50 +69,66 @@ options require the `pytest-pep8 plugin
+Instead of calling ``build_ext --inplace``, you can also install the package
+in editable mode::
+
+    pip install -e .[test]
```
Isn't this the preferred way? If so, this should be mentioned first before the `build_ext` option.

astrofrog replied:
I've switched this around

pllim commented on docs/development/testguide.rst:
```diff
@@ -143,63 +161,69 @@ Astropy Test Function
-The test suite can be run directly from the native ``pytest`` command. In this
-case, it is important for developers to be aware that they must manually
-rebuild any extensions by running ``setup.py build_ext`` before testing.
+Finally, the most robust way to run the tests (which can also be the slowest) is
```
Again, the preferred way should be mentioned first, or at least, link to this section from the `pytest` section above saying something like "but go to this section for the recommended way to run tests" or something like that.

astrofrog replied:
I moved tox to be first and edited the text accordingly

pllim commented on docs/install.rst:
```diff
@@ -13,7 +13,7 @@ Requirements
 - `Numpy`_ |minimum_numpy_version| or later
 
-- `pytest`_ 3.1 or later
+- `pytest <https://docs.pytest.org/en/latest/>`_ 3.1 or later
```
Didn't you bump to `pytest` 4 in `pytest-astropy`?

astrofrog replied:
I just removed this line since pytest is no longer a required dependency - and we already mention below that pytest-astropy is a test dependency - I'd rather just say that and not have to keep the minimum version of pytest in sync between pytest-astropy and the astropy docs.

pllim commented on docs/install.rst:
```diff
@@ -129,7 +132,11 @@ available optional dependencies, you can do::
+astropy, but if you are using a very recent version of Python, if a new version
+of astropy has just been released, or if you are building astropy for a platform
+that is not common, astropy will be installed from a source file. Note that in
+this case you will need a C compiler (e.g. ``gcc`` or ``clang``) to be installed
```
```suggestion
this case you will need a C compiler (e.g., ``gcc`` or ``clang``) to be installed
```

pllim commented on setup.cfg:
```diff
@@ -22,10 +20,12 @@ classifiers =
 zip_safe = False
 tests_require = pytest-astropy
-setup_requires = numpy>=1.16
+setup_requires = setuptools_scm
```
Is `setup_requires` still necessary?
```suggestion
```

astrofrog replied:
We could remove it, but leaving it allows commands like ``python setup.py --version`` to still work, which can be useful in some cases.

pllim replied:
I didn't even know `setup.py --version` existed. Okay.

pllim commented on setup.cfg:
```diff
@@ -46,9 +46,8 @@ asdf_extensions =
-    pytest-astropy
+    pytest-astropy>=0.8
     pytest-xdist
-    pytest-mpl
```
`pytest-mpl` is not part of `pytest-astropy`, so should we remove it from here?

astrofrog replied:
I think the reason I didn't include it for now is because if we leave it here it means we can never actually run the test suite without matplotlib - I need to go and make pytest-mpl not require matplotlib before we can add it back (at which point we'll just add it to pytest-astropy).

pllim replied:
Okay, I re-opened astropy/pytest-astropy#4 so we can follow-up there.

pllim commented on setup.cfg:
```diff
@@ -149,3 +130,42 @@ multi_line_output = 0
 include_trailing_comma = false
 length_sort_stdlib = true
+
+[coverage:run]
```
Ah, interesting. So you got rid of `coveragerc` file altogether.

astrofrog replied:
Indeed! Thanks @Cadair for letting me know this is possible :)

pllim commented on setup.py:
```diff
@@ -2,34 +2,78 @@
+""".lstrip()
+
+setup(use_scm_version={'write_to': os.path.join('astropy', 'version.py'),
+                       'write_to_template': VERSION_TEMPLATE},
```
I don't remember the template part from the helpers guide I read in Socorro. Why is this needed now?

astrofrog replied:
This was a recent development in the discussion above - basically this template allows the version number to automatically update when there are changes in the repository without having to re-run ``pip`` every time (otherwise the version number is frozen at each ``pip`` call).

mwcraig commented on .travis.yml:
```diff
@@ -204,31 +181,33 @@ before_install:
 
 install:
-    - if [[ $USE_CI_HELPERS == True ]]; then
+    - if [[ $TRAVIS_OS_NAME == osx || $TRAVIS_OS_NAME == windows ]]; then
```
Travis provides a usable python on linux but not windows or mac

astrofrog replied:
It's the opposite - it's not needed on Linux because then you can just set up Python with ``language: python``, but that isn't an option for Windows or MacOS X so we need to use conda there.

astrofrog commented on MANIFEST.in:
```diff
@@ -5,10 +5,11 @@ include CHANGES.rst
 include ah_bootstrap.py
 include setup.cfg
-include astropy/tests/coveragerc
+include .coveragerc
```
This is peripherally related to APE 17 although indeed not mentioned there - in standard Python packaging, the coveragerc should either be in ``.coveragerc`` or in ``setup.cfg`` - the fact we included it as part of the package itself was related to the astropy-helpers test runner. I will add a mention of this change to the migration guide.

astrofrog commented on setup.cfg:
```diff
@@ -46,9 +46,8 @@ asdf_extensions =
 [options.extras_require]
 test =
-    pytest-astropy
+    pytest-astropy>=0.8
```
I think it would make more sense to make this change in pytest-astropy, so that other packages benefit from this? Can you make a PR there?

pllim replied:
Okay, please see astropy/pytest-astropy#30

astrofrog commented on docs/conf.py:
```diff
@@ -106,11 +105,7 @@
-# The full version, including alpha/beta/rc tags.
-release = astropy.__version__
-
+release = version = astropy.__version__
```
Done

astrofrog commented on docs/development/astropy-package-template.rst:
```diff
@@ -35,34 +35,37 @@ by ``CHANGES.md`` in the instructions.
 
 #. Update the version number in ``setup.cfg`` to the version you're about to
-   release, without the ``.dev`` suffix (e.g. ``0.1``).
+   release, without the ``.dev`` suffix (e.g. ``0.1``). If your package uses
```
I agree with this in principle but I'd like to delay this a little just at least until all our own coordinated packages have migrated.

astrofrog commented:
@pllim - I think whenever branches are fetched the tags will be pulled automatically so there should be no need to do this. I suggest we leave it as-is for now, and if people do run into issues with unpulled tags (which would surprise me) we can add a note?

astrofrog commented:
@pllim - ok I added a mention of --tags in relevant ``git fetch`` commands

astrofrog commented:
@mhvk - since you reviewed this extensively before, are you happy with the current state?

astrofrog commented on .travis.yml:
```diff
@@ -15,65 +15,39 @@ os:
 
-# The apt packages below are needed for sphinx builds, which can no longer
-# be installed with sudo apt-get.
 # tzdata is included to ensure system leap seconds are up to date.
```
Fixed!

astrofrog commented on docs/development/ccython.rst:
```diff
@@ -34,22 +32,15 @@ Using Numpy C headers
+    import numpy
 
     def get_extensions():
         return Extension(name='myextension', sources=['myext.pyx'],
```
Fixed!

astrofrog commented:
@mhvk - thanks for the remaining comments! I decided to just go ahead and fix these since I agree and I might forget to do it if I don't do it now :)

pllim commented:
@astrofrog , can you please rebase and undo #9878 ? Thank you!

astrofrog commented:
@pllim - done!

mhvk commented:
My sense would be to get this in. Any objections?

Cadair commented:
Do it! I have two follow-on branches already :stuck_out_tongue: :shipit: :ship:

taldcroft commented:
@bsipocz @eteq - are you each OK with this PR?

pllim commented:
My vote is to merge now. I think @bsipocz already said she got no problem with merging this. And as for @eteq , he can open follow-up PR if he is unhappy with the content when he is available again. 😉 :shipit:

taldcroft commented:
OK then, I'm good with merging!

@astrofrog had previously asked the coordination committee about process for approval of this PR, so I'm speaking now with my CoCo hat on.

pllim commented:
Merged. Thank you, @astrofrog and everyone involved! 🎉

Cadair commented:
:rocket: :rocket: :tada: :fireworks: :dancing_women: :dancing_men:

mhvk commented:
Thanks, indeed, @astrofrog and everyone else!

astrofrog commented:
Thanks for merging this! And please don't hesitate to ping me or @Cadair if you run into any issues locally or in future pull requests :pray:

namurphy commented:
Thank you to everyone who made this PR happen!  I'd also like to thank everyone who has contributed to astropy-helpers over the last several years, as it has been really helpful to the community during the time that it has been around.

mhvk commented:
Seconded! I've been convinced over this PR how great it is to get rid of astropy helpers, but it made things much easier when these new nifty tools weren't around!

eteq commented:
I had been planning to review this but @pllim is definitely right that I can always do my own PR with any changes, because big-picture I was already fine with it!
